### PR TITLE
Moving release process to Bintray

### LINKED ISSUE #70: Refactor the release process to be based on Bintray
Similarly to https://github.com/mapbox/mapbox-gl-native/pull/13999.

### PR DESCRIPTION
**Pull request by langsmith**
Resolves #70 by moving the release/snapshot process to Bintray.

Related: 

- https://github.com/mapbox/mapbox-plugins-android/pull/961
- https://github.com/mapbox/mapbox-java/pull/1037

### CODE REVIEW

**Comment by langsmith** [comment]
More weird Bintray 🐛 at the finish line of these Bintray prs. For this repo, it's a `Invalid publication 'MapboxGesturesPublication': artifact file does not exist: '/Users/LangstonSmith/Mapbox/mapbox-gestures-android/library/build/libs/library-sources.jar.asc'` message

On my local machine, `make release` works just fine for building the project. When I try to publish to Bintray via `make publish-stable`, I get the following 💥below:

Screen Shot 2019-07-01 at 10 18 27 AM

No worries if not @LukasPaczos , but any ideas about the missing `.asc` file?

**Comment by langsmith** [comment]
Found https://github.com/gradle/gradle/issues/3635 and tried some suggestions in that thread, but nothing worked.

**Comment by langsmith** [comment]
Also related:

https://github.com/asciidoctor/asciidoctorj/issues/470

https://discuss.gradle.org/t/how-to-publish-artifacts-signatures-asc-files-using-maven-publish-plugin/7422/23

https://github.com/bintray/gradle-bintray-plugin/issues/255#issuecomment-423678744 looks promising but that's with the assumption that the `.asc` file is even generated.

**Comment by LukasPaczos** [comment]
> any ideas about the missing .asc file

I believe that those are the signed artifacts. Are you sure that your `mapboxGpgPassphrase` was correctly set on your local machine when you tested?

**Comment by langsmith** [comment]
Figured out what the issue was. I hadn't deleted the `mvn-push-android.gradle` file and the reference to it at https://github.com/mapbox/mapbox-gestures-android/blob/master/build.gradle#L28 🤦‍♂🤦‍♂🤦‍♂🤦‍♂. It was interfering with my new Bintray setup .

When I deleted it, ran `make release`, and then ran `make publish-stable`, everything uploaded to Bintray just fine 🎊  https://bintray.com/mapbox/mapbox/mapbox-android-gestures

_As a test_ and to get MavenCentral sync permission, I've published 0.5.0 on Bintray. We can delete later.

Screen Shot 2019-07-02 at 10 24 59 AM

Still getting that weird message when I request SNAPSHOT, but we can cross that bridge when we get to it. Things seemed to resolve themselves in https://github.com/mapbox/mapbox-java/pull/1037
Screen Shot 2019-07-02 at 10 28 23 AM

Screen Shot 2019-07-02 at 10 28 55 AM

**Comment by langsmith** [comment]
Looks like we're good to go with SNAPSHOTs too

Screen Shot 2019-07-02 at 10 49 42 PM

File: Makefile [hunk 1]
```diff
@@ -11,9 +11,13 @@ test:
 release:
 	./gradlew :library:assembleRelease
 
-publish:
-	export IS_LOCAL_DEVELOPMENT=false; ./gradlew :library:uploadArchives
+publish-stable:
+	export IS_LOCAL_DEVELOPMENT=false; ./gradlew :library:bintrayUpload
 
 publish-local:
```
Comment (LukasPaczos) [inline Makefile]: This does not work anymore and can be removed.

File: build.gradle [hunk 2]
```diff
@@ -5,10 +5,14 @@ buildscript {
     repositories {
         google()
         jcenter()
+        // For publishing files to Bintray
+        maven { url 'https://oss.jfrog.org/artifactory/oss-snapshot-local/' }
```
Comment (LukasPaczos) [inline build.gradle]: This line is not necessary to publish to arttifactory. Could you remove it?

File: circle.yml [hunk 3]
```diff
@@ -62,23 +62,27 @@ jobs:
...
       - run:
           name: Build libraries
           command: make release
       - deploy:
-          name: Publish to Maven Central
+          name: Publish to Bintray
           command: |
-            if [ "${CIRCLE_BRANCH}" == "master" ]; then
-            make publish ;
+            if [[ $CIRCLE_BRANCH == master ]] || [[ $CIRCLE_BRANCH == release-v* ]] || [[ $CIRCLE_TAG == v* ]]; then
```
Comment (LukasPaczos) [inline circle.yml]: We are using the `v[number]` convention in this repo, so `[[ $CIRCLE_BRANCH == release-v* ]]` can be skipped.

File: gradle/javadoc.gradle [hunk 4]
```diff
@@ -0,0 +1,19 @@
+apply from: "${rootDir}/gradle/dependencies.gradle"
+
+android.libraryVariants.all { variant ->
```
Comment (LukasPaczos) [inline gradle/javadoc.gradle]: This file means that `/library/javadoc.gradle` can be removed.

File: Makefile (hunk 1 above)
Comment (langsmith) [reply Makefile]: Removed

File: build.gradle (hunk 2 above)
Comment (langsmith) [reply build.gradle]: Removed

File: circle.yml (hunk 3 above)
Comment (langsmith) [reply circle.yml]: Removed

File: gradle/javadoc.gradle (hunk 4 above)
Comment (langsmith) [reply gradle/javadoc.gradle]: Removed

**Comment by langsmith** [comment]
>The release process is missing the step that changes the version name and removes the SNAPSHOT reference in order to support a full tag release. Check out this step.

Added to `circle.yml`.
>Also, the readme should be updated with the new snapshot repository URL.

Updated.

**Comment by langsmith** [comment]
Ok @LukasPaczos , just pushed a commit to address your latest review.

File: Makefile [hunk 5]
```diff
@@ -11,9 +11,9 @@ test:
 release:
 	./gradlew :library:assembleRelease
 
-publish:
-	export IS_LOCAL_DEVELOPMENT=false; ./gradlew :library:uploadArchives
+publish-stable:
+	export IS_LOCAL_DEVELOPMENT=false; ./gradlew :library:bintrayUpload
```
Comment (LukasPaczos) [inline Makefile]: `IS_LOCAL_DEVELOPMENT` is not used.

File: library/build.gradle [hunk 6]
```diff
@@ -1,4 +1,8 @@
 apply plugin: 'com.android.library'
+apply plugin: 'com.jfrog.bintray'
```
Comment (LukasPaczos) [inline library/build.gradle]: Already applied in the `gradle-bintray.gradle`.

File: library/build.gradle [hunk 7]
```diff
@@ -1,4 +1,8 @@
...
+apply plugin: "com.jfrog.artifactory"
```
Comment (LukasPaczos) [inline library/build.gradle]: Same as above.

File: library/build.gradle [hunk 8]
```diff
@@ -1,4 +1,8 @@
...
+apply plugin: 'digital.wup.android-maven-publish'
```
Comment (LukasPaczos) [inline library/build.gradle]: Same as above.

File: library/build.gradle [hunk 9]
```diff
@@ -43,4 +47,5 @@ dependencies {
     androidTestImplementation dependenciesList.testEspressoCore
 }
 
-apply from: 'javadoc.gradle'
\ No newline at end of file
+apply from: "${rootDir}/gradle/gradle-bintray.gradle"
+apply from: "${rootDir}/gradle/artifact-settings.gradle"
```
Comment (LukasPaczos) [inline library/build.gradle]: Same as above.

File: Makefile (hunk 5 above)
Comment (langsmith) [reply Makefile]: Removed

File: library/build.gradle (hunk 7 above)
Comment (langsmith) [reply library/build.gradle]: Removed

File: library/build.gradle (hunk 9 above)
Comment (langsmith) [reply library/build.gradle]: Removed

File: library/build.gradle (hunk 8 above)
Comment (langsmith) [reply library/build.gradle]: Removed

**Pull request merged**